python-api/livy/client_builder.py

```diff
@@ -0,0 +1,95 @@
+#
+# Licensed to Cloudera, Inc. under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  Cloudera, Inc. licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from ConfigParser import SafeConfigParser
+from StringIO import StringIO
+from client_factory import _ClientFactory
+from urlparse import urlparse
+import os
+
+"""A builder for Livy clients."""
+class ClientBuilder(object):
+
+    _CONFIG_SECTION = 'env'
+    _LIVY_CLIENT_CONF_DIR = "LIVY_CLIENT_CONF_DIR"
+
+    def __init__(self, uri, load_defaults=True, conf_dict = dict()):
+        """
+        Creates a new builder that will load the config
+        :param uri: Livy server uri generated from urlparse lib
+        :type uri: ParseResult generated from urlparse lib
+
+        >>> from urlparse import urlparse
+        >>> uri = urlparse("http://test")
+
+        :param boolean load_defaults: Default value is true. If true, loads the
+            livy client configuration from the files 'livy-client.conf' and
+            'spark-defaults.conf'. The path to the files should be defined in the
+            'LIVY_CLIENT_CONF_DIR' environment variable
+        :param dict: Default value is empty dict. The key-value pairs in the dict
+            will be loaded to the config
+        :returns An instance of the ClientBuilder
+        """
+        self._config = SafeConfigParser()
+        self._load_config(load_defaults, conf_dict)
+        self._set_uri(uri)
+
+    def _set_uri(self, uri):
+        self._config.set(self._CONFIG_SECTION, 'livy.uri', uri.geturl())
```

Comment:
If you don't return `self` then this doesn't really look like a builder. Not sure if that pattern is common in python, though.